# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

In [ ]:
# Lambda functions allow us to use functions without saving them to our server
# Used in case we want to just use a function quickly
# Sort of like "list comprehension of functions" because it can be done in one line

#### Syntax

In [4]:
# This is a regular function
def addTwo(num):
    return num + 2
print(addTwo(4))

#Lambda syntax for defining a function without saving it to a variable
# Always starts with lambda keyword and returns what we want, then put our argument after it in ()
print((lambda num: num + 2)(4))


6
6


#### Saving to a Variable

In [6]:
# We can save lambdas as variables
# If we save lambdas as a variable, we can call it like a function
# It returns the result (very similar to a function) unless you print it

f_test = lambda num: num + 2
f_test(4)
print(f_test(4))

6


#### Multiple Inputs

In [8]:
# Multiple inputs with no variable
print((lambda x, y, z: x * y * z)(3, 5, 8))

# Multiple inputs saved to a variable
num_test = lambda x, y, z: x * y * z
print(num_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [9]:
def multiple(func, num):
    """
    func expects a lambda function
    and num expects an integer
    """
    return func(num)

print(multiple(lambda x: x * x, 4))
# note the syntax of it -- num is inside the parentheses because it's an argument to the multiple function

16


#### Returning a Lambda from a Function

In [15]:
# Regular way of defining a function
def multiply_test(num):
    return num * 4
print(multiply_test(2))


# Syntax for writing functions inside functions
def return_func():
    test = 4
    def multiply(num):
        return num * 4
    return multiply        # we need to make sure we return inside both functions otherwise code won't work

f_return = return_func()
print(f_return(2))


# We like to use lambda functions instead of using functions within functions which can get hairy
def return_lamb(a, b):
    return lambda c, d: c + d + a + b

r_return = return_lamb(3, 4)    # 3 & 4 are our a & b
print(r_return(5,6))            # 5 & 6 are our c & d

8
8
18


In [18]:
def my_func(num1, num2):
    print(f"Num 1 & Num 2 are: {num1}, {num2}")
    return lambda num3, num4: (num3 * num4) + (num1 * num2)


# This uses lambdas saved to a variable
alex_var = my_func(1, 2)    # 1 & 2 are num1 & num2 respectively
print(alex_var(3, 4))       # 3 & 4 are num3 & num4 respectively


# This uses lambdas without saving to a variable
print(my_func(1,2)(3,4))

Num 1 & Num 2 are: 1, 2
14
Num 1 & Num 2 are: 1, 2
14


#### If Statements within Lambdas

In [19]:
# Syntax for if statements in lambdas
# Lambda x: True if (condition) else False

f_conditional = lambda num: num * 2 if num > 10 else num + 2
print(f_conditional(8))
print(f_conditional(10))
print(f_conditional(12))

10
12
24


In [24]:
# Example of if, elif, else using lambdas

                                                            # below is our if statement
alex_example = lambda num: f"{num} is less than 5" if num < 5 else\
                                                            # below is our elif statement
(f"{num} is greater than 5 & less than 10" if num < 10 else\
                                                            # below is our else statement
f"{num} is greater than 10")

print(alex_example(3))
print(alex_example(8))
print(alex_example(10))

3 is less than 5
8 is greater than 5 & less than 10
10 is greater than 10


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [25]:
f = lambda num: num ** 3

print(f(2))
print(f(3))

8
27


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

In [ ]:
# Map is similar to a for loop

#### Syntax

In [27]:
# Syntax for utilizing map function
# map(function, iterable)
# Normally the usage of map happens with pre-defined user functions, but we can use lambdas

# Utitlizing a map function with a regular pre-defined user function
def squared(num1, num2):
    if num1 < 10 and num2 < 10:
        return num1**2, num2**2
    else:
        return num1, num2
    
nums = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_nums = list(map(squared, nums, more_nums))  # this is saying for each num in nums, and for each num in more_nums
print(squared_nums)

# the number 20 in list nums did not run because more_nums reached the end of the list (has one less index)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [31]:
# Syntax for map & lambda functions
# list(map(lambda x: (transformation on x), iterable))
# This usually happens in one line so it's cleaner code

squared_nums_again = list(map(lambda num1, num2: (num1**2, num2**2) if num1 < 10 and num2 <10\
                             else (num1, num2), nums, more_nums))
print(squared_nums_again)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [33]:
num_list = [5, 10, 13, 21]

doubled_and_minus = list(map(lambda num: num*2 - 1, num_list))
print(doubled_and_minus)

[9, 19, 25, 41]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [42]:
# Syntax for filter
# list(filter(function, iterable))

names_list = ["chris", "steve", "perry", "kelsey"]

def y_names(names):
    if names[-1].lower() == "y":
        return True
    else:
        return False    # so if the name does not end with 'y', it will be filtered out of the list
    
new_names_list = list(filter(y_names, names_list))
print(new_names_list)
print(names_list)

['perry', 'kelsey']
['chris', 'steve', 'perry', 'kelsey']


#### Using Lambda's with Filter()

In [44]:
new_names_lamb = list(filter(lambda name: True if name[-1] == 'y' else False, names_list))
print(new_names_lamb)

['perry', 'kelsey']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [63]:
from statistics import mean

num_list = [3, 4, 5, 8, 10]

above_mean_list = list(filter(lambda num: True if num < mean(num_list) else False, num_list))
print(mean(num_list))
print(above_mean_list)

6
[3, 4, 5]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [66]:
# Snytax for reduce
# reduce(function, iterable)
# Reduction is a mathematical technique that uses folding to reduce it to a single cumulative value

from functools import reduce
list_1 = [2, 4, 6, 8, 10]

# Using a user-defined function with reduce
def add_nums (num1, num2):
    return num1 + num2

result_add = reduce(add_nums, list_1)
print(result_add)

# Using a for loop instead
result = 0
for num in list_1:
    result += num
print(result)

30
30


#### Using Lambda's with Reduce()

In [68]:
reduce_lamb = reduce(lambda num1, num2: num1+num2, list_1)
print(reduce_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [71]:
from functools import reduce
num_list = [3, 7, 10, 16]

result = 1
for num in num_list:
    result*= num
print(result)

reduced_multiples = reduce(lambda num1, num2: num1*num2, num_list)
print(reduced_multiples)

3360
3360


## Recursion <br>
<p>Recursion means that a function is calling itself, so it constantly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

##### Implementing a Base Case

In [76]:
# Anything recursive can be done iteratively (ie without recursion)

def add_nums(num):
    # first we want to set our base case
    if num <= 1:
        print("add_nums(1) = 1")
        return num
    else:
        print(f"add_nums({num}) = {num} + add_nums({num-1})")
        return num + add_nums(num-1)

print(add_nums(5))

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1
15


#### Writing a Factorial Function

In [79]:
# 5! = 5 * 4 * 3 * 2 * 1

def factorial(num):
    # set our base case
    if num <=1:
        return 1
    else:
        return num * factorial(num-1)
    
print(factorial(5))

120
120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [12]:
# return num - sub_nums(num - 1)
# Recursion is like a tree. You keep going down the tree until you reach the base case (roots)
# Once you've reached the roots, you can start growing the tree

def sub_nums(num):
    if num <= 1:
        print("sub_names(1) = 1")
        return num
    else:
        print(f"sub_names({num}) = {num} - sub_nums({num-1})")
        return num - sub_nums(num-1)
print(sub_nums(5))

#sub_nums(1) = 1
#sub_nums(2) = 2 - sub_nums(1) = 2 - 1 = 1
#sub_nums(3) = 3 - sub_nums(2) = 3 - 1 = 2
#sub_nums(4) = 4 - sub_nums(3) = 4 - 2 = 2
#sub_nums(5) = 5 - sub_nums(4) = 5 - 2 = 3

sub_names(5) = 5 - sub_nums(4)
sub_names(4) = 4 - sub_nums(3)
sub_names(3) = 3 - sub_nums(2)
sub_names(2) = 2 - sub_nums(1)
sub_names(1) = 1
3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [6]:
# Generators are like functions that have 'return' values that we can iterate through by "saving" the return value
# If we create a regular function, it only creates one return value
# and once the function is done, it 'resets' it's return value so it doesn't save the last return value

def my_range(stop, start, step = 2):
    while start < stop:
        yield start # yield keyword is what makes it a generator -- we yield values, we do not return values
        start += step
        
for i in my_range(stop=20, start=2):
    my_generator_value = i
    print(my_generator_value)

2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [8]:
# bad, never create infinite loops
def bad_range(stop, start, step=1):
    while start < stop:
        yield start
        # this is why we need to increment our generators to avoid infinite while loops!
        # start += step
        
for i in bad_range(10, 1):
    print(i)
    # if we forget to increment the generator, then it will continuously just print out i = 1

1
2
3
4
5
6
7
8
9


#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [27]:
def squared_gen(num):
    while num >= 0:
        yield num ** 2
        num -= 1
        
print([i for i in squared_gen(10)])


[100, 81, 64, 49, 36, 25, 16, 9, 4, 1, 0]


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [39]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

places_without_spaces = list(filter(lambda word: True if word.strip() != "" else False, places))
print(places_without_spaces)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [37]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]


author.sort(key = lambda full_name: full_name.lower().split()[-1])
print(author)



# the key parameter in sort() is an optional parameter that helps to specify the sorting criteria
# .split() splits a string (at the spaces by default) and returns a list
# remember the syntax for lambda is lambda keyword: whatever we want to return (aka our last name)
# so basically we are using split() to split the full name into a list containing the first/middle/last name
# to access the last name, we can just index to the last item in a list, which we know we can do by using [-1]

Carter
aNisimov
Garfield
hassELHOFF
Bernstein
['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [36]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

celsius_places = list(map(lambda city_temp_tup: (city_temp_tup[0], (city_temp_tup[1] * (9/5) + 32)), places))
print(celsius_places)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [6]:
def fib(num):
    # set our base case
    if num <= 1:
        return 1
    else:
        print(f"fib({num}) = fib({num-2}) + fib({num-1})")
        return (fib(num-2) + fib(num-1))
    
print(fib(5))

# fib(0) = 1
# fib(1) = 1
# fib(2) = fib(0) + fib(1) = 1 + 1 = 2
# fib(3) = fib(1) + fib(2) = 1 + 2 = 3
# fib(4) = fib(2) + fib(3) = 2 + 3 = 5
# fib(5) = fib(3) + fib(4) = 3 + 5 = 8

fib(5) = fib(3) + fib(4)
fib(3) = fib(1) + fib(2)
fib(2) = fib(0) + fib(1)
fib(4) = fib(2) + fib(3)
fib(2) = fib(0) + fib(1)
fib(3) = fib(1) + fib(2)
fib(2) = fib(0) + fib(1)
8
